### PR TITLE
[Gemma4] LM head + on-device sampling

### LINKED ISSUE #44953: [Gemma4] LM head + on-device sampling
# [Gemma4] LM head + on‑device sampling

Parent: #44962

## Motivation

`vocab_size = 262 144` (per‑device shard 32 768 on 1x8). The LM head matmul + sampling sit at the tail of every decode step and are difficult to parallelize with the next prefill / decode. We need to confirm:

- The `lm_head` matmul program config is tuned for `[B, 1, hidden] × [hidden, V_per_dev]` at `B = 32`.
- On‑device sampling (init message `On-device sampling initialized (vocab=262144, per_device=32768)`) actually fires for batched decode and isn't falling back to host argmax.
- `tt_transformers/tt/generator.py:_get_sampling_contract` returns a `sampling_dp > 1` config when running DP=4.

## Scope

- Audit `lm_head` matmul program config; sweep on Blackhole + Wormhole.
- Confirm on‑device sampling kernel covers top‑k / top‑p / greedy code paths.
- Add `test_lm_head` PCC test if missing for vocab=262 144.
- Profile sampling step in tracy — should be < 5 % of decode step.

## Acceptance criteria

- `test_lm_head` and `test_sampling` PCC ≥ 0.999.
- On‑device sampling enabled for batch=32 decode without falling back to host argmax.
- LM head + sampling dispatch ≤ 10 % of decode step time.

## Files

- `models/demos/gemma4/tt/model.py` (`lm_head`, sampling integration)
- `models/demos/gemma4/tests/unit/test_lm_head.py`
- `models/demos/gemma4/tests/unit/test_sampling.py`
- `models/tt_transformers/tt/generator.py` (`_get_sampling_contract`)

### PR DESCRIPTION
**Pull request by cleonidouTT**
### PR Category
Performance

### Summary
Implements the Gemma4 bringup work tracked in issue #44953 for the `gemma-4-31B-it` variant on T3K (1×8). At the tail of every decode step the LM head matmul + sampling are on the critical path and were either untuned or running on the host argmax fallback; this PR closes the gap end-to-end.

- **LM head matmul tuned for `[B=32, 1, 5376] × [5376, V/TP]`.** Added `_get_lm_head_program_config` in `models/demos/gemma4/tt/model.py` that picks `MatmulMultiCoreReuseMultiCast1DProgramConfig` from the live tensor shapes and the device's `compute_with_storage_grid_size()` (8×8 on WH, 8×10 on BH). For 31B this resolves to `K_tiles=168`, `N_tiles=1024`, `in0_block_w=8`. Wired into `Gemma4Model.__call__` for both decode and prefill-last-token paths.
- **On-device sampling enabled for batched decode.** `Gemma4Model._supports_on_device_sampling` is now bound to `self.sampling is not None` rather than hardcoded to `False`. `SamplingGenerator` is constructed for any `tp > 1` mesh where the per-device padded vocab fits in 64 K. Tracy confirms `SamplingDeviceOperation` + `TopKDeviceOperation` dispatch — no host-side full-vocab argmax.
- **`sampling_dp` propagated for the row-shard contract.** `Gemma4Model` now sets `self.sampling_dp = mesh_device.shape[0]` and `_make_sampling_args` propagates the same value, so `tt_transformers/tt/generator.py:_get_sampling_contract` can return `sampling_dp > 1` on a DP=4 / 4×8 mesh. Preserves existing 1×N behavior (`sampling_dp == 1`).
- **PCC coverage at 31B shapes.** Added `test_lm_head_decode_batch32` (`[1, 1, 32, hidden]` × LM head, PCC ≥ 0.999) and four `test_sampling_*` tests (greedy/top-k/top-p constrained + log-probs PCC) against torch reference logits at vocab=262 144. `pcc_thresholds.json` updated with `gemma-4-31B-it` entries for `1x8` and `1x4`. Decorators include `(1, 4)` so the 4-chip Blackhole quietbox can run them today.
- **Tracy scopes for the perf gate.** Added `gemma4_lm_head` and `gemma4_sampling` `signpost` markers around the LM head matmul (+ softcap) and `self.sampling.sample(...)` in `Gemma4Model.__call__`, gated on `is_decode` so prefill last-token calls don't contaminate the decode totals. `op_perf_results.py --signpost ...` can carve those regions out of a `text_demo` decode trace.
- **Test collection hardening.** `test_factory.py:is_moe_model()` now tolerates `AutoConfig.from_pretrained` raising on unrecognized model types (older `transformers` versions), so pytest collection no longer aborts on `gemma4`.

Acceptance criteria for #44953:
- `test_lm_head` / `test_sampling` PCC ≥ 0.999 — added and passing on 1×4 BH (CI also runs on 1×8 WH/BH).
- On-device sampling fires at batch=32 — verified by Tracy: `SamplingDeviceOperation` + `TopKDeviceOperation` dispatch, no host argmax.
- LM head + sampling ≤ 10% of decode step — isolated 31B-shape measurements on 1×4 BH: LM head matmul ~1.83 ms, on-device sampling ~10.2 ms (TopK-dominated). Projecting per-device shard halving on T3K (TP=8 vs 4) and using the README's published 105 ms/decode-step for 31B, this satisfies the combined ≤10% gate (~6 %) and is borderline at the sampling-only <5 % gate (~5 %). The final acceptance run is `pytest models/demos/gemma4/demo/text_demo.py::test_demo -k '1x8 and prefill_128'` under Tracy on T3K; this PR ships the signposts that make that measurement a one-liner.

Closes #44953

### Notes for reviewers
- Primary focus areas:
  - `models/demos/gemma4/tt/model.py` — `_get_lm_head_program_config`, `_make_sampling_args`, `self.sampling_dp` plumbing, signpost placement (decode-only guard).
  - `models/demos/gemma4/tests/unit/test_lm_head.py` — new `test_lm_head_decode_batch32` (uses the same program config helper as the model path).
  - `models/demos/gemma4/tests/unit/test_sampling.py` — four new tests + a pure-Python regression test for `sampling_dp` against a fake 4×8 mesh.
- Trade-offs and explicit non-goals:
  - LM head still uses **DRAM TILE weights + 1D mcast** (LLMs guide §4.3.2.3), not DRAM-sharded weights + L1 width-sharded activations (§4.3.2.2). If T3K Tracy shows the LM head dominating after this PR, that's the next perf lever — explicitly out of scope here.
  - `TopK` at vocab=262 144 is the dominant cost in `gemma4_sampling`. If the strict <5 % sampling-only gate must be tightened, optimizing or chunking the distributed top-K is the lever; not addressed in this PR.
  - `sampling_dp > 1` has code-complete plumbing + a pure-Python regression test but no on-hardware validation since Galaxy 4×8 isn't available and is still 🔴 in the README support matrix. Once 4×8 wiring lands this will exercise immediately.
- 1×4 vs 1×8 in tests: `test_lm_head_decode_batch32` and the sampling tests are parametrized for both `(1, 4)` and `(1, 8)`. `test_sampling_log_probs_pcc` is skipped on 1×4 because `LogProbsCalculator._is_supported()` in `models/common/sampling/tt_log_probs.py` hardcodes device counts to `{8, 32}` — same constraint other models hit, called out via `pytest.skip` rather than silently passing.
- The 31B-only PCC thresholds and Tracy targets are owned by this PR; smaller variants (E2B / E4B / 26B-A4B) continue to use shape-derived configs without regressions (no per-variant overrides changed).

### CI Status
_Auto-generated on every push. Badges update live. Click a badge to filter runs by this branch._

- [](https://github.com/tenstorrent/tt-metal/actions/workflows/sanity-tests.yaml?query=branch:cleonidouTT/gemma4_lmhead_on-device-sampling)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/runtime-sanity-tests.yaml?query=branch:cleonidouTT/gemma4_lmhead_on-device-sampling)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml?query=branch:cleonidouTT/gemma4_lmhead_on-device-sampling)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/tt-metal-l2-nightly.yaml?query=branch:cleonidouTT/gemma4_lmhead_on-device-sampling)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/all-model-tests.yaml?query=branch:cleonidouTT/gemma4_lmhead_on-device-sampling)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select.yaml?query=branch:cleonidouTT/gemma4_lmhead_on-device-sampling)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-t3k.yaml?query=branch:cleonidouTT/gemma4_lmhead_on-device-sampling)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-galaxy.yaml?query=branch:cleonidouTT/gemma4_lmhead_on-device-sampling)

### CODE REVIEW

File: models/demos/gemma4/tt/model.py [hunk 1]
```diff
@@ -415,6 +474,13 @@ def __init__(
...
 
+        # sampling_dp: number of independent sampling groups (one per mesh row).
+        # This is 1 for standard TP-only meshes (e.g. 1x8), and >1 for multi-row
+        # meshes where each row samples users independently (e.g. Galaxy 4x8).
+        #
+        # tt_transformers' Generator reads this attribute via _get_sampling_contract.
+        self.sampling_dp = mesh_device.shape[0] if is_mesh else 1
+
         # On-device sampling (greedy/top-k/top-p) — avoids reading full vocab logits to CPU
         self.sampling = None
```
Comment (handrewsTT) [inline models/demos/gemma4/tt/model.py:505]: Does this not disable on device sampling?

Comment (cleonidouTT) [reply models/demos/gemma4/tt/model.py:505]: `sampling_dp = 1` means "a single sampling group," not "sampling disabled." The thing that would disable it is `self.sampling` being `None` (i.e. `tp == 1`, or per-device vocab > 64K)

Comment (handrewsTT) [reply models/demos/gemma4/tt/model.py:505]: no my comment is specifically about self.sampling=None still being set here? line 485

Comment (cleonidouTT) [reply models/demos/gemma4/tt/model.py:505]: `self.sampling = None` is just the **default initialization** of the attribute, not a disable. The very next `if` overwrites it with a real `SamplingGenerator` whenever the mesh supports on-device sampling.

```485:498:models/demos/gemma4/tt/model.py
        self.sampling = None
        if is_mesh and tp > 1:
            per_device_padded = _compute_per_device_vocab(hf_config.vocab_size, tp)
            if per_device_padded <= 64 * 1024:
                self.sampling = SamplingGenerator(...)
                ...
        self._supports_on_device_sampling = self.sampling is not None
```

So after `__init__`:
- **1×8 / 1×4 TP mesh, vocab fits** → `self.sampling` gets reassigned to a real `SamplingGenerator`, so it is **not** `None`, and `_supports_on_device_sampling = True`.
- `self.sampling` only *stays* `None` in the genuinely-unsupported cases: In those cases on-device sampling legitimately can't run and the Generator falls back to host argmax.

The reason it must be initialized first is that the `if`/nested-`if` can both be false (e.g. `tp == 1`), and line 498 (`self.sampling is not None`) plus the Generator's `_get_sampling_contract` (`getattr(..., "sampling", None)`) need the attribute to always exist. Without it, those paths would hit an `AttributeError` on the unsupported configs.

Comment (handrewsTT) [reply models/demos/gemma4/tt/model.py:505]: ah I see, that codeblock was collapsed out for me. Thanks for the response

**Review by handrewsTT — APPROVED** [review]

**Review by mtairum — APPROVED** [review]

**Comment by cleonidouTT** [comment]
vLLM CI: [vLLM nightly tests](https://github.com/tenstorrent/tt-metal/actions/runs/28371500962)
Metal CI: [All Model Tests — Tiers[1,2] Tests[unit,e2e] Model[gemma-4] SKUs[all]](https://github.com/tenstorrent/tt-metal/actions/runs/28185735288)

**Comment by cleonidouTT** [comment]
Branche's Latest commit: [da8e335](https://github.com/tenstorrent/tt-metal/pull/45324/commits/da8e335d9e3ffb971f04878615dbd05b0c147b32)
Metal CI run branch commit: [0886eff](https://github.com/tenstorrent/tt-metal/commit/0886eff830e75c5760c2146241143bc4c63d5c0b) (https://github.com/tenstorrent/tt-metal/actions/runs/28185735288)

**Review by arginugaTT — APPROVED** [review]
